File: Javascript/Arrays/Problem18.js

 // Brute Force Approch



let nums=[-2,1,-3,4,-1,2,1,-5,4];

let max=0;
for(let i=0;i<nums.length;i++){
  let sum=0;
  for(let j=i;j<nums.length;j++){
    sum+=nums[j];
   
   if(sum>max){
     max=sum;
   }

  }
  
}

console.log(max)


// Kadane Algorithm


let max=-Infinity;
let sum=0;
for(let i=0;i<nums.length;i++){
    sum+=nums[i];

    if(sum>max){
        max=sum;
    }
    else if(sum<0){
        sum=0;
    }
}

console.log(sum);

File: C++/Arrays/Problem18.cpp


// 1. Brute Force Approch


#include <iostream>
#include <bits/stdc++.h>
using namespace std;

int main() 
{
    int n=9;
    int arr[n]={-2,1,-3,4,-1,2,1,-5,4};
    int max=INT_MIN;
    for(int i=0;i<n;i++){
      int sum=0;
      for(int j=i;j<n;j++){
        sum+=arr[j];
        if(sum>max){
          max=sum;
        }
      }
    }
    std::cout << max << std::endl;
    return 0;
}


// 2  . Optimal Solution 


class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        
        
        int max=INT_MIN;
        int sum=0;
        for(int i=0;i<nums.size();i++){
          sum+=nums[i];
          if(sum>max){
            max=sum;
          }
          if(sum<0){
            sum=0;
          }
        }
        return max;
    }
};